[[im/subscribe]]Add IM subscription support upon IM read

yunhanw-google opened the pull request:
#### Problems:
Follow IM spec chapter 5 and IM encoding spec 7.3 and 7.4 to achieve
subscription functionality.
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Interaction-Model.adoc#5-subscribe-interaction
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#subscriberequestmessage

#### Change overview
- Establishing distinct subscriptions and tracking their respective configurations from multiple clients simultaneously.
- Sharing of reporting logic between reads and subscriptions to maximize re-use.
- Multiple paths per subscription
- Basic chunking of data across multiple reports.
- Automatic liveness assessment on client and server and termination of subscription accordingly.
- Eventing support in addition to attribute reporting.
- Spec compliance to the encoding specification with TLV-based payloads.
- Preservation of the application level APIs for client-side reporting, but with a swap of the internals to use the new machinery

#### Testing
-- Add unit test for positive and negative subscribe test
-- Add integrated cirque test for subscribe.

erjiaqing commented on src/app/ReadClient.h:
```diff
@@ -88,11 +96,14 @@ class ReadClient : public Messaging::ExchangeDelegate
+        SubscriptionIdle,          ///< The client is maintaining subscription
     };
 
+    bool IsValidSubscription(uint64_t & aSubscriptionId) { return aSubscriptionId == mSubscriptionId; }
```
Consider add `mRoleId == Subscribe`

yunhanw-google replied:
Done, thanks

erjiaqing commented on src/app/ReadClient.cpp:
```diff
@@ -464,5 +521,154 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+    CancelLivenessCheckTimer();
+    ChipLogProgress(DataManagement, "Refresh LivenessCheckTime with %d seconds", mFinalSyncIntervalSeconds);
+    err = InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionMgr()->SystemLayer()->StartTimer(
+        mFinalSyncIntervalSeconds * 1000, OnLivenessTimeoutCallback, this);
```
Consider kMillisecondsPerSecond

yunhanw-google replied:
Done, thanks

erjiaqing commented on src/app/ReadHandler.h:
```diff
@@ -163,6 +178,10 @@ class ReadHandler : public Messaging::ExchangeDelegate
     Messaging::ExchangeManager * mpExchangeMgr = nullptr;
     InteractionModelDelegate * mpDelegate      = nullptr;
     bool mInitialReport                        = false;
```
Can we remove this flag by checking the state?

yunhanw-google replied:
i am hesitated on this point. currently read and subscribe share reporting and reportable state, I use mInitialReport to mark the difference. if using state for checking, i have to introduce more state, like initial report, initial reportable, later subscribeReport, subscribeReportable. 
Since this code is already there, we may revisit it later. thanks

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -109,6 +116,30 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
             }
+            else
+            {
+                for (auto & dirtyPath : mDirtyPathList)
```
What if we have multiple dirtyPath?

AttributeDataElement::Builder will be reused, and will cause incorrect state.

yunhanw-google replied:
Thanks for catching up this issue, fixed, thanks

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -109,6 +116,30 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
+            {
+                for (auto & dirtyPath : mDirtyPathList)
+                {
+                    if (dirtyPath.IsDirty())
```
dirtyPath.IsDirty seems weird.

yunhanw-google replied:
rename dirtypath to path.

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,62 @@ void Engine::Run()
+            while (clusterInstance != nullptr)
+            {
+                bool setSuccess = false;
+                // If dirty path intersects with interested path from handler, it mark this in dirty path list and also
```
For now, we need a additional check for the existence of the path.
For subscription, we should not report non-exist path.

yunhanw-google replied:
agree. would do it in seperate pr

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -317,6 +355,9 @@ CHIP_ERROR Engine::ScheduleRun()
 void Engine::Run()
 {
+#if !CHIP_SYSTEM_CONFIG_NO_LOCKING
+    ScopedLock lock(*this);
```
Why do we need lock here since this function should be called in main loop.

yunhanw-google replied:
SetDirty may be called from different thread. Jerry, what do you think of?

yunhanw-google replied:
We may create the seperate PR for lock?

andy31415 replied:
We can create a separate PR, however that should be in master or a separate commit on top of this one.

erjiaqing commented on src/app/reporting/Engine.h:
```diff
@@ -102,6 +104,17 @@ class Engine
     static void Run(System::Layer * aSystemLayer, void * apAppState);
 
+#if !CHIP_SYSTEM_CONFIG_NO_LOCKING
+    class ScopedLock
```
Is it possible to create a seperate PR?

We can try:

class ScopedLock {
public:
ScopedLock(Mutex & mutex) : mMutex(mutex) { mutex.Lock(); }
~ScopedLock(Mutex & mutex)  { mutex.Unlock(); }
private:
Mutex & mMutex;
}

yunhanw-google replied:
i can create the seperate one

yunhanw-google replied:
removed,

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,78 @@ void Engine::Run()
+        {
+            ClusterInfo * clusterInstance = readHandler.GetAttributeClusterInfolist();
+
+            while (clusterInstance != nullptr)
```
Why not for loop here?

yunhanw-google replied:
that is linked list, while loop is better? why for loop is better?

erjiaqing replied:
For loop is safer, what if some one added a "continue" in the loop in the future?

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,78 @@ void Engine::Run()
+                }
+                else
+                {
+                    ChipLogDetail(DataManagement, "<RE> SetDirty skip this path since no subscriber interested in it");
```
This comment seems misplaced/

yunhanw-google replied:
yep, updated, thanks

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,78 @@ void Engine::Run()
+                            path.SetDirty();
+                            clusterInstance->SetDirty();
+                            ChipLogDetail(DataManagement, "<RE> SetDiry Success");
+                            return;
```
What if we have more than one subscribed interested in the same path? Do we need to mark all of them dirty?

yunhanw-google replied:
the overlaped path has been removed here, https://github.com/project-chip/connectedhomeip/blob/master/src/app/ReadHandler.cpp#L286

erjiaqing replied:
Consider the case:

Node A subscribes Path a/b/c
Node B subscribes Path a/b/c

Now, call set dirty with a/b/c, what will happed to Node A and Node B's subscription sessions.

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,78 @@ void Engine::Run()
+                if (clusterInstance->IsAttributePathSupersetOf(aClusterInfo) ||
+                    aClusterInfo.IsAttributePathSupersetOf(*clusterInstance))
+                {
+                    for (auto & path : mDirtyPathList)
```
Can we move this outside the very outter loop?
i.e.
1. Get a seat in dirty path list
2. Iterate all client and mark necessary path dirty.
3. If we marked any path dirty, set that item dirty.

yunhanw-google replied:
the intuitive loop
-- Loop all handle, 
-- then loop corresponding interested clusters for each handler
--then check it against the dirty path list
it seems it is more suitable here.

andy31415 commented on src/app/InteractionModelDelegate.h:
```diff
@@ -195,6 +196,16 @@ class InteractionModelDelegate
+    /**
+     * Notification that a Subscribe Response has been processed and application can do further work .
+     */
+    virtual CHIP_ERROR SubscribeResponseProcessed(const ReadClient * apReadClient) { return CHIP_ERROR_NOT_IMPLEMENTED; }
```
why the difference between  NOT_IMPLEMENTED here and NO_ERROR_below?
should we/could we make some of these pure virtual? I imagine if we say 'NOT IMPLEMENTED' we may as well make the methods '=0' and force implementation.

yunhanw-google replied:
yep, should be CHIP_ERROR_NOT_IMPLEMENTED. agree, I would file seperate PR to address this

asakhart replied:
what is the reason for not '=0'? Won't compile in some other place?

andy31415 commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,36 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+CHIP_ERROR InteractionModelEngine::OnReportData(Messaging::ExchangeContext * apExchangeContext, const PacketHeader & aPacketHeader,
+                                                const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err          = CHIP_NO_ERROR;
```
remove this variable and use it only in the for loop without a break. We should avoid C-style variable declarations (where everything goes on top)

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,36 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+{
+    CHIP_ERROR err          = CHIP_NO_ERROR;
+    uint64_t subscriptionId = 0;
+    ReportData::Parser report;
```
please separate declarations by usage instead of placing everything at the top of the method. The  order of declarations  seems to be: create  a reader out of the payload, create a report out of the reader. This seems clearer:

```cpp

System::PacketBufferTLVReader reader;

reader.Init(aPayload.Retain()); 
ReturnLogErrorOnFailure(reader.Next());   /// is this needed? Next returns a CHIP_ERROR

ReportData::Parser report;

ReturnLogErrorOnFailure(report.Init(reader));

uint64_t subscriptionId = 0;
￼ReturnLogErrorOnFailure(report.GetSubscriptionId(&subscriptionId));
￼
/// ...
```

yunhanw-google replied:
agree, done, thanks

andy31415 commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,36 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+    ReportData::Parser report;
+    System::PacketBufferTLVReader reader;
+    reader.Init(aPayload.Retain());
+    reader.Next();
```
note the above: Next returns a CHIP_ERROR

yunhanw-google replied:
done, thanks

andy31415 commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,36 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+
+    for (auto & readClient : mReadClients)
+    {
+        if (readClient.IsSubscriptionIdle())
```
this is deeply nested. Could we reduce using continue? 

```cpp
for (auto & readClient : mReadClients) 
{
    if (readClient.IsSubscriptionIdle()) {
        continue;
    }
    if (readClient.IsValidSubscription(subscriptionId)) {
       continue;
    }
    
    
    // ...
}

```

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,36 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+                readClient.SetExchangeContext(apExchangeContext);
+                err = readClient.ProcessReportData(std::move(aPayload));
+                readClient.ClearExchangeContext();
+                apExchangeContext = nullptr;
```
why is this setting to null necessary?

instead of break below we can just return it seems.

yunhanw-google replied:
done, thanks

andy31415 commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,36 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+            if (readClient.IsValidSubscription(subscriptionId))
+            {
+                readClient.SetExchangeContext(apExchangeContext);
+                err = readClient.ProcessReportData(std::move(aPayload));
```
report and reader above do not seem to be used anymore, yet it did a 'retain' of the payload.
Would it make sense to have a method of 'get the payload back from the report' instead of using retain & move?

TLVPacketBufferBackingStore::Release() seems to exist and returns the underlying packet buffer.

yunhanw-google replied:
The PacketBufferHandle has been refactored, the Retain will return an new handle to the packet puffer it wrapped, which will do a extra release after the handle get out of scope, so don't need to call Release manually after Retain

PacketBufferHandle only have retain api, and don't have release api
In TestSystemPacketBuffer.cpp, I see it use retain in 27 places, but no manual release there

andy31415 commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -300,12 +331,22 @@ CHIP_ERROR InteractionModelEngine::OnMessageReceived(Messaging::ExchangeContext
     else if (aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::ReadRequest))
     {
-        err = OnReadInitialRequest(apExchangeContext, aPacketHeader, aPayloadHeader, std::move(aPayload));
+        err =
```
these all should be `return` however I can see hyow we are trying to do a minimal  change here

yunhanw-google replied:
yep, i would create one pr before this pr to cleanup this, thanks

andy31415 commented on src/app/InteractionModelEngine.h:
```diff
@@ -111,6 +111,14 @@ class InteractionModelEngine : public Messaging::ExchangeDelegate
 
+    /**
+     *  Creates a new read client and send SubscribeRequest message to the node using the read client. User should use this method
+     * since it takes care of the life cycle of ReadClient.
```
"User should use this method" - what does this mean? Use as opposed to what other alternative.

Consider phrasing as "Handles lifecycle of the ReadClient, as opposed to <what is the altenative>".
Also define what the 'lifecycle' is. I do not understand what the  readclient lifecycle is here and none of the return value or parameters say aything about a 'read client'.

yunhanw-google replied:
updated, please take a look, thanks

andy31415 commented on src/app/ReadClient.cpp:
```diff
@@ -58,13 +63,20 @@ void ReadClient::Shutdown()
     }
+    if (IsSubscription())
+    {
+        mFinalSyncIntervalSeconds = 0;
```
could we/sound we move these out of the if  condition? It seems that we want everything cleared in general, except IsSubscription also cancels a liveness check.

It is easyer to reason about a 'this method always sets this value to 0/null' than a 'this is set to 0 only if a condition is met'

I assume here that we  do not require 'must only set these to 0 if we are a subscription'

yunhanw-google replied:
agree, done thanks

andy31415 commented on src/app/ReadClient.cpp:
```diff
@@ -189,6 +205,23 @@ CHIP_ERROR ReadClient::SendStatusReport(CHIP_ERROR aError)
+        if (IsAwaitingInitialReport())
+        {
+            MoveToState(ClientState::AwaitingSubscribeResponse);
+            expectResponse = true;
```
expectResponse seems to be set to true IFF status is AwaitingSubscribeResponse

Could we drop this bool and only use the status?

yunhanw-google replied:
agree, done, thanks

andy31415 commented on src/app/ReadClient.cpp:
```diff
@@ -464,5 +521,154 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    CancelLivenessCheckTimer();
+    ChipLogProgress(DataManagement, "Refresh LivenessCheckTime with %d seconds", mFinalSyncIntervalSeconds);
```
how often will this be called? Are there concerns about spammyness of these messages? naming seems to suggest logs frequency on O(seconds)

yunhanw-google replied:
as long as subscription is alive, we would periodically check this every X seconds.

andy31415 commented on src/app/ReadClient.cpp:
```diff
@@ -464,5 +521,154 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+
+CHIP_ERROR ReadClient::RefreshLivenessCheckTimer()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
please  declare variables when used: drop this like and say: 

```cpp
CHIP_ERROR err = InteractionModelEngine::.....;
```

below.

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/ReadClient.cpp:
```diff
@@ -464,5 +521,154 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+void ReadClient::OnLivenessTimeoutCallback(System::Layer * apSystemLayer, void * apAppState)
+{
+    ReadClient * const client = reinterpret_cast<ReadClient *>(apAppState);
+    ChipLogError(DataManagement, "Subscription Liveness timeout, would shutdown");
```
s/would shutdown/shutting down/

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/ReadClient.cpp:
```diff
@@ -464,5 +521,154 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+    if (client->IsFree())
+    {
+        ChipLogError(DataManagement,
+                     "ReadClient::OnLivenessTimeoutCallback invoked on a free client! This should be a bug in CHIP stack!");
```
s/should/is/

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/ReadClient.cpp:
```diff
@@ -464,5 +521,154 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+    client->ShutdownInternal(CHIP_ERROR_TIMEOUT);
+}
+
+CHIP_ERROR ReadClient::ProcessSubscribeResponse(System::PacketBufferHandle && aPayload)
```
Not for this PR, but thinking aloud:

I see this System::PacketBufferHandle passed around everywhere, however because it is a systempacketbufferhandle and we move, it is passed by reference and we periodically need to do std::move or Retain().

Do we really need a systempacketbuffer? It seems to me that we need a 'buffer with length' for all of these. Passing in a Span<> seems better: no more std::move and it decouples compilation from a more complex type (packetbufferhandle has length management and various fancy methods) to a basic 'sized buffer'.

Thoughts? Should we/could we improve our API internally even  on ToT?

mrjerryjohns replied:
I think it's a fair point to switch to a leaner construct here internally within ReadClient. But how much would we actually save? Let's do some back-of-envelope checks...

* `InteractionModelEngine::OnMessageReceived`: Today, the InteractionModelEngine is just a forwarder of the incoming packet and 'routes' it to the right client/handler object to service it. It today, doesn't make assumptions about the subsequent management of that packet, or the lifetime of it after that. If we go down this Span model, it would mean that the IMEngine is explicitly going to toss out that packet buffer after each message has been processed. This is a strong assumption to make, but one I think that is consistent for all IM processing today (i.e no one holds on to the packet buffer for longer than the call to process the packet).

That would save 6 calls to the respective handling objects in this function.

* In most of the handling objects (CommandSender, CommandHandler, ReadClient, etc) that are themselves ExchangeDelegates, they will be provided a PacketBuffer as part of `OnMessageReceived`. In most cases, they just internally position a Reader on that and don't pass around packet buffer handles after that. It's _only_ ReadClient and ReadHandler that do this modal selection between processing read vs. subscribe messages.

So that's two APIs each X 2. So 4.

Total = 10 functions?

Worth it?

mrjerryjohns replied:
At any rate, I recommend we tackle this later.

andy31415 commented on src/app/ReadClient.cpp:
```diff
@@ -464,5 +521,154 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+    SuccessOrExit(err);
+
+    err = subscribeResponse.GetFinalSyncIntervalSeconds(&mFinalSyncIntervalSeconds);
+    if (CHIP_END_OF_TLV == err)
```
if err is END_OF_TLV what will the value of mFinalSyncIntervalSeconds? same for above for getting subscriptionId.

My understanding of the code (as a reader) is: err will be END_OF_TLV if we encounter the end of the TLV buffer _before_ we find the corresponding value. In this case:


if I fail to find a subscriptionid, we get the default 0. Why is this ok?
If I fail to find a interval seconds, I will leave 'mFinalSyncIntervalSeconds' to 0 (after searching through comments ... my initial expectation is that if TLV is not found the value should not be defined...0 or unchanged both seem find as an implementation when you get an error). Why is this ok and not an error?

yunhanw-google replied:
per spec, subscribe id and final sync internal are mandatory, you are right, here they are bugs, just fix them, thanks

andy31415 commented on src/app/ReadClient.h:
```diff
@@ -88,11 +96,14 @@ class ReadClient : public Messaging::ExchangeDelegate
+        SubscriptionIdle,          ///< The client is maintaining subscription
     };
 
+    bool IsValidSubscription(uint64_t & aSubscriptionId) { return aSubscriptionId == mSubscriptionId && mRoleId == RoleId::Subscribe; }
```
Why do we pass the argument by reference? 

I wonder if 'Valid' is the best name. I would expect this name to be 'validation' which would check something (like non-negative numbers or maybe not 0 or something like that).

Insteade this checks if the client is a subscription for the given id. How about

```cpp
bool IsClientForSubscription(uint64_t subscription);
```

yunhanw-google replied:
agree, done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -56,6 +60,15 @@ CHIP_ERROR ReadHandler::Init(Messaging::ExchangeManager * apExchangeMgr, Interac
 {
+    if (IsSubscription())
+    {
+        mSubscriptionId     = 0;
```
I suggest to move everything except CancelTimer outside of the if: it is easier to reason about saying  'shutdown ensures these values are 0'.

yunhanw-google replied:
agree, done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -108,8 +128,33 @@ CHIP_ERROR ReadHandler::OnStatusReport(Messaging::ExchangeContext * apExchangeCo
     switch (mState)
     {
+    case HandlerState::SubscribingResponding:
+        ClearInitialReport();
```
why is this a method? this seems the only usage and it is internal and all it does is 'mInitialReport = false;'

Could we simplify code?

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -446,5 +513,100 @@ void ReadHandler::MoveToNextScheduledDirtyPriority()
+    InteractionModelEngine::GetInstance()->GetReportingEngine().ScheduleRun();
+}
+
+CHIP_ERROR ReadHandler::RefreshSubscribeSyncTimer(void)
```
use () instead of (void)

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -108,8 +128,33 @@ CHIP_ERROR ReadHandler::OnStatusReport(Messaging::ExchangeContext * apExchangeCo
     {
+    case HandlerState::SubscribingResponding:
+        ClearInitialReport();
+        RefreshSubscribeSyncTimer();
```
RefreshSubscribeSyncTimer returns a CHIP_ERROR. please do not silently ignore (at least log)

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -128,16 +173,35 @@ CHIP_ERROR ReadHandler::OnStatusReport(Messaging::ExchangeContext * apExchangeCo
 
 CHIP_ERROR ReadHandler::SendReportData(System::PacketBufferHandle && aPayload)
 {
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
used in only one place, please declare when used instead of  at the top of the function.

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -128,16 +173,35 @@ CHIP_ERROR ReadHandler::OnStatusReport(Messaging::ExchangeContext * apExchangeCo
     MoveToState(HandlerState::Reporting);
-    return mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
-                                      Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
```
above we move to state Reporting and move back to reportable on failure.

Why has this not been done in previous code?
Would it make sense to instead move to Reporting only on success since the previous state should be 'Reportable' as we do a VerifyOrReturn?

yunhanw-google replied:
just remove that reportable. thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -446,5 +513,100 @@ void ReadHandler::MoveToNextScheduledDirtyPriority()
+
+CHIP_ERROR ReadHandler::SendSubscribeResponse()
+{
+    SubscribeResponse::Builder response;
```
separate declarations please:

```cpp
 System::PacketBufferHandle packet = 
     System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
￼VerifyOrReturnLogError(!packet.IsNull(), CHIP_ERROR_NO_MEMORY);

 System::PacketBufferTLVWriter writer;
 writer.Init(std::move(packet));

 SubscribeResponse::Builder response;
￼ReturnLogErrorOnFailure(response.Init(&writer));
 // ...
```

yunhanw-google replied:
done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -446,5 +513,100 @@ void ReadHandler::MoveToNextScheduledDirtyPriority()
+    ReturnLogErrorOnFailure(mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::SubscribeResponse, std::move(packet),
+                                                       Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse)));
+
+    return CHIP_NO_ERROR;
```
end with `return mpExchangeCtx->SendMessage(....)` instead of using a macro and returning CHIP_NO_ERROR

yunhanw-google replied:
done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -446,5 +513,100 @@ void ReadHandler::MoveToNextScheduledDirtyPriority()
+
+CHIP_ERROR ReadHandler::ProcessSubscribeRequest(System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
```
separate out variables to be closer to initialization/usage. All variables on top of the method is C, not C++.

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/ReadHandler.cpp:
```diff
@@ -446,5 +513,100 @@ void ReadHandler::MoveToNextScheduledDirtyPriority()
+    ReturnLogErrorOnFailure(subscribeRequestParser.GetMaxIntervalSeconds(&mMaxIntervalSeconds));
+
+    // TODO: Use GetSecureRandomData to generate subscription id
+    // err = Platform::Security::GetSecureRandomData((uint8_t *) &mSubscriptionId, sizeof(mSubscriptionId));
```
can we do this now? rand() is really unsafe.

yunhanw-google replied:
it needs #include <support/crypto/CHIPRNG.h>, but somehow CHIPRNG.h is missing

yunhanw-google replied:
just use GetRandU64 from RandUtils.h, although it is not still safe, but it is better than rand. would follow up with Evgeni about CHIPRNG.h

andy31415 replied:
RandUtils should be ok - we will have a separate task to make randutils use the DRNG that spec decides (i.e. it will be secure)

yunhanw-google replied:
sure, thanks

andy31415 commented on src/app/ReadHandler.h:
```diff
@@ -117,18 +124,26 @@ class ReadHandler : public Messaging::ExchangeDelegate
     };
 
+    static void OnRefreshSubscribeTimerSyncCallback(System::Layer * apSystemLayer, void * apAppState);
+    CHIP_ERROR RefreshSubscribeSyncTimer(void);
```
use () instead of (void)

yunhanw-google replied:
Done, thanks

andy31415 commented on src/app/reporting/Engine.h:
```diff
@@ -102,6 +104,17 @@ class Engine
+    class ScopedLock
+    {
+    public:
+        ScopedLock(Engine & aEngine) : mEngine(aEngine) { mEngine.mAccessLock.Lock(); }
```
scoped locks should be on mutexes not other types

yunhanw-google replied:
removed, thanks

yunhanw-google replied:
removed, thanks

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -35,6 +35,13 @@ CHIP_ERROR Engine::Init()
     mNumReportsInFlight  = 0;
     mCurReadHandlerIdx   = 0;
+#if !CHIP_SYSTEM_CONFIG_NO_LOCKING
+    CHIP_ERROR err = System::Mutex::Init(mAccessLock);
```
why doesn't mAccessLock have an ::Init method instead?

yunhanw-google replied:
removed, thanks

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -53,10 +60,11 @@ EventNumber Engine::CountEvents(ReadHandler * apReadHandler, EventNumber * apIni
 {
     CHIP_ERROR err                              = CHIP_NO_ERROR;
-    AttributePath::Builder attributePathBuilder = aAttributeDataElementBuilder.CreateAttributePathBuilder();
+    AttributeDataElement::Builder attributeDataElementBuilder = aAttributeDataList.CreateAttributeDataElementBuilder();
```
throughout his codebase: `goto exit` forces us to place these declarations on top. In this case we declare them 20 lines before they are actually used :( this makes me very unhappy as a reviewer, needing to always look up and down for scope.

Nothing we can consistently do, however venting some frustration and asking to try hard to stop with the goto exit approach.

yunhanw-google replied:
agree.

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -100,15 +108,38 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
+            {
+                for (auto & path : mDirtyPathList)
+                {
+                    if (path.IsDirty())
```
why is this check here? The member names is 'mDirtyPathlist' so I imagine the content being 'dirty paths'. Is this  untrue and should we change the member name? or not do the dirty check?

yunhanw-google replied:
update it with mChangeSetPathList, true, mDirtyPathlist is confusing name.

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -100,15 +108,38 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
+                        }
+                        else if (path.IsAttributePathSupersetOf(*clusterInfo))
+                        {
+                            err = RetrieveClusterData(attributeDataList, *clusterInfo);
```
does this mean that if one element changes, we report all elements? I.e. if I ask for EP 1, CL 2 all attributes, then if a single attribute changes I get all attribute values? Should this not be filtered by path in some way?

yunhanw-google replied:
1. if someone mutate one element(path)in server and that element is part of interested cluster path(clusterInfo), then we just report that particular element change, this code is "clusterInfo->IsAttributePathSupersetOf(path)", 

2. if someone mutate one root change(path), which is super set of small interested cluster path, then we only report that small interested cluster path, this code is "path.IsAttributePathSupersetOf(*clusterInfo“

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -100,15 +108,38 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
+                        }
+                        else
+                        {
+                            continue;
```
is superset the only option? It may be worth mentioning in a comment: partial overlap is not possible, hence the 'continue' here: clusterInfo and path have nothing in common.

yunhanw-google replied:
done, thanks

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -244,6 +275,13 @@ CHIP_ERROR Engine::BuildAndSendSingleReportData(ReadHandler * apReadHandler)
+    if (apReadHandler->IsSubscription())
+    {
+        uint64_t subscriptionId = 0;
+        apReadHandler->GetSubscriptionId(subscriptionId);
```
does GetSubscriptionId return a CHIP_ERROR? if so, it should be handled.

yunhanw-google replied:
no error, thanks

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,78 @@ void Engine::Run()
     }
+    for (auto & path : mDirtyPathList)
+    {
+        path.ClearDirty();
```
this makes mDirtyPathList NOT have the dirty flag anymore, so the name is misleading. Please rename to something that represents what it contains (it is not the 'list of dirty paths' because here they are obviously not dirty anymore)

yunhanw-google replied:
yep, rename it to mChangeSetPathList, thanks

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,78 @@ void Engine::Run()
+    {
+        path.ClearDirty();
+    }
+    return;
```
remove this return

yunhanw-google replied:
done, thanks

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,78 @@ void Engine::Run()
+{
+    InteractionModelEngine * imEngine = InteractionModelEngine::GetInstance();
+#if !CHIP_SYSTEM_CONFIG_NO_LOCKING
+    ScopedLock lock(*this);
```
similar  to question above: why are locks needed? I believe we are moving to a model where the locking is app defined and we assume a single event loop from CHIP perspective.

What is the threading that makes life difficult here? Why is this needed for SetDirty?

yunhanw-google replied:
ok, removed the lock

andy31415 commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +377,78 @@ void Engine::Run()
+
+            while (clusterInstance != nullptr)
+            {
+                bool setSuccess = false;
```
I see setSuccess being set to true and then a `return` so this may never be set to tue. Is this correct? Does this have any use?

yunhanw-google replied:
yep, agree, just removed, thanks

andy31415 commented on src/app/tests/TestReadInteraction.cpp:
```diff
@@ -187,13 +190,34 @@ CHIP_ERROR ReadSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVWriter * ap
-    err = apWriter->Put(TLV::ContextTag(AttributeDataElement::kCsTag_Data), kTestFieldValue1);
-    SuccessOrExit(err);
-    err = apWriter->Put(TLV::ContextTag(AttributeDataElement::kCsTag_DataVersion), version);
+        if (apWriter != nullptr)
```
if apWriter is nullptr, the setting of *apDataExists will be reset/lost below in the =true statement. Is this a bug?

yunhanw-google replied:
fixed, thanks

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -346,11 +435,6 @@ CHIP_ERROR Engine::SendReport(ReadHandler * apReadHandler, System::PacketBufferH
-
-    if (err != CHIP_NO_ERROR)
-    {
-        mNumReportsInFlight--;
```
Can we move `mNumReportsInFlight++` to `apReadHandler->SendReportData`?

yunhanw-google replied:
mNumReportsInFlight is part of reporting engine

erjiaqing commented on src/app/tests/TestReadInteraction.cpp:
```diff
@@ -165,39 +168,58 @@ class MockInteractionModelApp : public chip::app::InteractionModelDelegate
 
 namespace chip {
 namespace app {
 CHIP_ERROR ReadSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVWriter * apWriter, bool * apDataExists)
```
Should be

```
CHIP_ERROR ReadSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVWriter * apWriter, bool * apDataExists)
{
    uint64_t version = 0;
    ChipLogDetail(DataManagement, "TEST Cluster %" PRIx32 ", Field %" PRIx32 " is dirty", aClusterInfo.mClusterId,
                  aClusterInfo.mFieldId);

    if (apDataExists != nullptr)
    {
        *apDataExists = (aClusterInfo.mClusterId == kTestClusterId && aClusterInfo.mEndpointId == kTestEndpointId);
    }

    if (apWriter == nullptr)
    {
        return CHIP_NO_ERROR;
    }

    if (!(aClusterInfo.mClusterId == kTestClusterId && aClusterInfo.mEndpointId == kTestEndpointId))
    {
        return apWriter->Put(chip::TLV::ContextTag(AttributeDataElement::kCsTag_Status),
                             chip::Protocols::InteractionModel::ProtocolCode::UnsupportedAttribute);
    }

    ReturnErrorOnFailure(apWriter->Put(TLV::ContextTag(AttributeDataElement::kCsTag_Data), kTestFieldValue1));
    return apWriter->Put(TLV::ContextTag(AttributeDataElement::kCsTag_DataVersion), version);
}
```

yunhanw-google replied:
would follow up this.

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +368,63 @@ void Engine::Run()
+            //If aClusterInfo is part of existing dirty path, skip it
+            //If aClusterInfo is super set of one existing path, update that path with new path.
+            if (path.IsAttributePathSupersetOf(aClusterInfo)) {
+                return;
```
Consider the following case:
- A path is marked dirty
- One new subscription comes
- (Report Engine run)

mrjerryjohns commented on src/app/InteractionModelDelegate.h:
```diff
@@ -196,13 +197,24 @@ class InteractionModelDelegate
+    /**
+     * Notification that Subscription has been established successfully and application can do further work in handler.
+     */
+    virtual CHIP_ERROR SubscriptionEstablished(const ReadHandler * apReadHandler) { return CHIP_ERROR_NOT_IMPLEMENTED; }
```
I don't see the difference between the `SubscriptionEstablished` callback and `SubscribeResponseProcessed` callbacks - what do clients do differently?

yunhanw-google replied:
SubscriptionEstablished is for read handler.

mrjerryjohns replied:
In that case, please name the above one `SubscriptionEstablished` as well. `SubscribeResponseProcessed` is leaking details of the messages in the interaction, when what clients care about is just the logical act of subscription establishment and its completion.

asakhart replied:
It's still confusing, reviewing the changes I won't know why both are there.

mrjerryjohns commented on src/app/InteractionModelDelegate.h:
```diff
@@ -196,13 +197,24 @@ class InteractionModelDelegate
+     */
+    virtual CHIP_ERROR SubscribeResponseProcessed(const ReadClient * apReadClient) { return CHIP_ERROR_NOT_IMPLEMENTED; }
+
+    //TODO: Refactor callback delegate into client and server delegate for different IM interaction
```
I've filed #9460.

yunhanw-google replied:
sure, thanks

mrjerryjohns commented on src/app/InteractionModelDelegate.h:
```diff
@@ -196,13 +197,24 @@ class InteractionModelDelegate
+    virtual CHIP_ERROR SubscriptionEstablished(const ReadHandler * apReadHandler) { return CHIP_ERROR_NOT_IMPLEMENTED; }
+
+    /**
+     * Notification that a read client has completed the current interaction.
```
This is specific to a read, and not a subscription right? Because for a subscription, it lives on indefinitely till either the client or server terminate the subscription.

Comment should be updated to indicate that this is only for _successful_ reads:

```suggestion
     * Notification that a read interaction was completed on the client successfully.
```

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/InteractionModelDelegate.h:
```diff
@@ -196,13 +197,24 @@ class InteractionModelDelegate
      * @retval # CHIP_ERROR_NOT_IMPLEMENTED if not implemented
      */
-    virtual CHIP_ERROR ReadDone(const ReadClient * apReadClient, CHIP_ERROR aError) { return CHIP_ERROR_NOT_IMPLEMENTED; }
+    virtual CHIP_ERROR ReadDone(const ReadClient * apReadClient) { return CHIP_ERROR_NOT_IMPLEMENTED; }
```
I can't comment by the line for `ReadError`, so commenting here.

We should add an error callback for Subscriptions? (i.e OnSubscriptionError())

Those should convey a couple of distinct errors (defined as their own error codes):
1. Subscription terminating on the client due to a timeout waiting for the sync.
2. Subscription terminating on the server due to a timeout waiting for the sync.
3. Subscription terminating on the server due to any number of reasons.

In the latter two cases, those errors get conveyed to the client through a status report. That status report's contents need to be delivered through the callback as well. The pattern I chose in the IM Prototype was to provide both a CHIP_ERROR status code + status report as a separate arg, and create a special status code called `CHIP_ERROR_STATUS_REPORT_RECEIVED` that can be used to figure out if a valid status report is provided in the second argument. Might be good to do that here.

yunhanw-google replied:
agree, we can mark this as todo?

mrjerryjohns commented on src/app/ReadClient.h:
```diff
@@ -52,6 +52,12 @@ namespace app {
 class ReadClient : public Messaging::ExchangeDelegate
 {
 public:
+    enum class RoleId : uint8_t
```
Minor nit: Perhaps instead of role, `InteractionType` would be better suited?

yunhanw-google replied:
agree, will fix it. thanks

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,42 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+
+        readClient.SetExchangeContext(apExchangeContext);
+        CHIP_ERROR err = readClient.ProcessReportData(std::move(aPayload));
+        readClient.ClearExchangeContext();
```
This should be a detail that is managed by the read client and is private to it. The IM engine does not have enough context to know when it should close or clear the EC that is actually now 'owned' by the ReadClient. This becomes especially true later on when we have multiple chunked reports, so it would be good to establish this pattern correctly now.

yunhanw-google replied:
Done, move it to reader client.

mrjerryjohns commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,42 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+        readClient.ClearExchangeContext();
+        if (err != CHIP_NO_ERROR)
+        {
+            readClient.Shutdown();
```
As per discussions before, we should move to a model where the allocation for client-side objects is owned by the application, and not done through a pool in the IM. Where possible, we should minimize the need to use such pools.

This does not have to happen in this PR, but worth filing an issue.

yunhanw-google replied:
Yep, plan to have the separate PR. Yufeng and Boris mentioned to use object pool, yufeng just finish the refactor this work, I would do it next week, thanks

mrjerryjohns commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +290,42 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
+    ReturnLogErrorOnFailure(report.Init(reader));
+
+    uint64_t subscriptionId = 0;
+    ReturnLogErrorOnFailure(report.GetSubscriptionId(&subscriptionId));
```
Doesn't `OnReportData` serve both repots coming back for reads and subscribes? If so, there is no subscription ID in reports coming back for reads right?

mrjerryjohns replied:
Same goes for a lot of the logic below...not sure how reads are working then...

yunhanw-google replied:
all read test has passed.
this is report received from Interaction Model Engine after subscription is established, which would use new exchange.
During the subscription establishment procedure, it uses same exchange, read client process all messaged using onMessageReceived.
After subscription established, handler would send report using new exchange, at that time, InteractionModelEngine would receive this report, set new exchange into read client, then read client can further send corresponding status report.

yunhanw-google replied:
just update the comment in code.

mrjerryjohns replied:
Got it. Would recommend re-naming this to `OnUnsolicitedReportData` to make it clear.

mrjerryjohns commented on src/app/ReadClient.cpp:
```diff
@@ -58,13 +63,24 @@ void ReadClient::Shutdown()
+        {
+            mpDelegate->ReadError(this, aError);
+        }
+        mpDelegate->ReadDone(this);
```
should be `else { mpDelegate->ReadDone(this); } ` right? Otherwise, ReadDone will called even in the error case?

yunhanw-google replied:
done, thanks

mrjerryjohns commented on src/app/ReadClient.cpp:
```diff
@@ -58,13 +63,24 @@ void ReadClient::Shutdown()
+    mSubscriptionId           = 0;
+    mRoleId = RoleId::Invalid;
+    mpExchangeMgr  = nullptr;
+    mpExchangeCtx  = nullptr;
```
Shouldn't we be closing the exchange here too in addition to nulling it out?

yunhanw-google replied:
i think the exchange layer would automatically do it.

yunhanw-google replied:
exchange manager would clear up these exchange context per previous boris and yufeng's refactor..

mrjerryjohns commented on src/app/ReadClient.cpp:
```diff
@@ -189,9 +208,25 @@ CHIP_ERROR ReadClient::SendStatusReport(CHIP_ERROR aError)
     System::PacketBufferHandle msgBuf = buf.Finalize();
     VerifyOrReturnLogError(!msgBuf.IsNull(), CHIP_ERROR_NO_MEMORY);
 
+    if (IsSubscription())
```
Since this PR doesn't implement chunking yet, in which case would we send a status report back for a read? It would only be for subscriptions right? So arguably, at the top of this function, we would have checked if this is of type subscription, and error out if it wasn't?

yunhanw-google replied:
currently, for read, client receives the non-empty report, and always send status report. since spec mentioned, "Data reports (i.e. non-empty ReportData messages) SHALL always require a response. In those cases, the SuppressResponse flag isn’t applicable and SHALL not be sent." in https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#reportdatamessage

mrjerryjohns commented on src/app/ReadClient.cpp:
```diff
@@ -319,10 +360,24 @@ CHIP_ERROR ReadClient::ProcessReportData(System::PacketBufferHandle && aPayload)
 
-    if (IsInitialReport())
+    err = report.GetSubscriptionId(&subscriptionId);
+    if (CHIP_NO_ERROR == err)
```
This isn't quite safe, because it makes it possible for a server to send back reports after a subscription has been established, without having a subscription ID in it, and that would cause this logic to assume that is ok.

yunhanw-google replied:
yep, agree, fixing it

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/ReadClient.cpp:
```diff
@@ -464,5 +515,146 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+                     "ReadClient::OnLivenessTimeoutCallback invoked on a free client! This is a bug in CHIP stack!");
+        return;
+    }
+    client->ShutdownInternal(CHIP_ERROR_TIMEOUT);
```
Would be good to add a more specific error here for liveness timeout failure to distinguish between other classes of timeouts (i.e response timeouts).

yunhanw-google replied:
yep, would add one todo here.

mrjerryjohns commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +292,44 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
 
+// During the subscription establishment procedure, it uses same exchange, read client process all messaged using onMessageReceived.
+// After subscription established, handler would send report using new exchange, at that time, InteractionModelEngine would
+// receive this report, set new exchange into read client, then read client can further send corresponding status report.
```
Suggested comment improvement:

"This function handles processing of un-solicited ReportData messages on the client, which can only occur post subscription establishment".

mrjerryjohns replied:
Also, please move function description comments to the header file.

yunhanw-google replied:
Done, thanks

yunhanw-google replied:
done, thnaks

mrjerryjohns commented on src/app/ReadClient.cpp:
```diff
@@ -459,5 +531,148 @@ CHIP_ERROR ReadClient::ProcessAttributeDataList(TLV::TLVReader & aAttributeDataL
+    return err;
+}
+
+CHIP_ERROR ReadClient::SendSubscribeRequest(ReadPrepareParams & aReadPrepareParams)
```
This is quite a bit duplicated with SendReadRequest. 

The core part of this function (from line 625 to 663) feels like it could be re-factored to a common function. 

To achieve that, you also need to re-factor your SubscribeRequest::Builder and ReadRequest::Builder to 'share' their common bits of attribute path, event path building. If re-factored into a 'ReadSubBase' that both extend from, that might make this possible.

yunhanw-google replied:
same thought on it. maybe we should add one todo here? or just file one ticket later

mrjerryjohns commented on src/app/ReadClient.h:
```diff
@@ -52,6 +52,12 @@ namespace app {
 public:
+    enum class InteractionType : uint8_t
+    {
+        Invalid = 0,
```
Do we need an invalid? What's the purpose?

yunhanw-google replied:
acutually no need, the default can be read.

mrjerryjohns commented on src/app/ReadClient.h:
```diff
@@ -74,8 +80,18 @@ class ReadClient : public Messaging::ExchangeDelegate
+    CHIP_ERROR SendSubscribeRequest(ReadPrepareParams & aSubscribePrepareParams);
     uint64_t GetAppIdentifier() const { return mAppIdentifier; }
     Messaging::ExchangeContext * GetExchangeContext() const { return mpExchangeCtx; }
+    bool IsSubscription() const { return mInteractionType == InteractionType::Subscribe; };
```
```suggestion
    bool IsSubscriptionType() const { return mInteractionType == InteractionType::Subscribe; };
```

Would recommend adding the other counterpart here too `IsReadType()` to be symmetric.

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/ReadClient.h:
```diff
@@ -84,11 +100,17 @@ class ReadClient : public Messaging::ExchangeDelegate
+        SubscriptionIdle,          ///< The client is maintaining subscription
     };
 
+    bool IsClientForSubscription(uint64_t aSubscriptionId)
```
Would recommend a less wordier version:

```suggestion
    bool IsMatchingClient(uint64_t aSubscriptionId)
```

yunhanw-google replied:
done, thanks

mrjerryjohns commented on src/app/ReadClient.h:
```diff
@@ -115,7 +138,10 @@ class ReadClient : public Messaging::ExchangeDelegate
      *
      */
     bool IsFree() const { return mState == ClientState::Uninitialized; }
+    bool IsSubscriptionIdle() const { return mState == ClientState::SubscriptionIdle; }
```
Can we shift these simple state queries to inside a State struct? That would reduce the # of methods here, and cleanly capture it within an encapsulated object.

yunhanw-google replied:
current state is     enum class ClientState, prefer seperate follow-up to do this, agree, it is better to have a state struct with these handy functions.

mrjerryjohns commented on src/app/ReadClient.h:
```diff
@@ -142,6 +171,9 @@ class ReadClient : public Messaging::ExchangeDelegate
     ClientState mState                         = ClientState::Uninitialized;
     uint64_t mAppIdentifier                    = 0;
     bool mInitialReport                        = true;
+    uint16_t mFinalSyncIntervalSeconds         = 0;
```
We should capture both min and max intervals, and not call it FinalSyncInterval (known issue with the encoding spec that will be fixed on Monday).

yunhanw-google replied:
Done, thanks

asakhart commented on src/app/InteractionModelDelegate.h:
```diff
@@ -196,7 +197,18 @@ class InteractionModelDelegate
+     */
+    virtual CHIP_ERROR SubscribeResponseProcessed(const ReadClient * apReadClient) { return CHIP_ERROR_NOT_IMPLEMENTED; }
+
+    // TODO: Refactor callback delegate into client and server delegate for different IM interaction
```
don't think this belongs here, for clarity, please focus on the task at hand to improve readability of these specific changes.

asakhart commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +292,44 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
     return err;
 }
 
+// During the subscription establishment procedure, it uses same exchange, read client process all messaged using onMessageReceived.
```
nit: ", same exchange is used. Read client processes all messages using onMessageReceived"

yunhanw-google replied:
Done, thanks

asakhart commented on src/app/InteractionModelEngine.cpp:
```diff
@@ -289,6 +292,44 @@ CHIP_ERROR InteractionModelEngine::OnWriteRequest(Messaging::ExchangeContext * a
 
+// During the subscription establishment procedure, it uses same exchange, read client process all messaged using onMessageReceived.
+// After subscription established, handler would send report using new exchange, at that time, InteractionModelEngine would
+// receive this report, set new exchange into read client, then read client can further send corresponding status report.
```
// After subscription is established, new exchange will be use to send further reports. InteractionModelEngine would receive this report, set new exchange into read client, then use it thereafter to send a corresponding status report.

yunhanw-google replied:
Done, thanks

asakhart commented on src/app/InteractionModelEngine.h:
```diff
@@ -111,6 +111,14 @@ class InteractionModelEngine : public Messaging::ExchangeDelegate
 
+    /**
+     *  Creates a new read client and send SubscribeRequest message to the node using the read client.
+     *  shutdown if fail to send it out
```
Creates a new read client and sends SubscribeRequest message to the node using the read client.
Shuts down on transmission failure.

yunhanw-google replied:
Done, thanks

asakhart commented on src/app/ReadClient.cpp:
```diff
@@ -24,25 +24,30 @@
+    mState                    = ClientState::Initialized;
+    mAppIdentifier            = aAppIdentifier;
+    mFinalSyncIntervalSeconds = 0;
+    mSubscriptionId           = 0;
```
Is Init() and Sutdown() called multiple times on the same read client?

yunhanw-google replied:
No, you cannot call multiple init without shutdwon for same read client, i think i will do the state check in the beginning, thanks

asakhart commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +367,65 @@ void Engine::Run()
+        if (readHandler.IsSubscription() && (readHandler.IsReportable() || readHandler.IsReporting()))
+        {
+            ClusterInfo * clusterInstance = readHandler.GetAttributeClusterInfolist();
+            while (clusterInstance != nullptr)
```
This is quite nested and hard to reason about, consider adding a helper function like SetReadHandlerDirty(readhandler)

instead of the block below.

ClusterInfo * clusterInstance = readHandler.GetAttributeClusterInfolist();
            while (clusterInstance != nullptr)

yunhanw-google replied:
just update it with SetChange, and update the implementation with 3 lines of code, thanks!

asakhart commented on src/app/tests/integration/chip_im_initiator.cpp:
```diff
@@ -48,9 +48,12 @@ constexpr size_t kMaxCommandMessageCount          = 3;
+constexpr uint32_t gMessageIntervalMsec   = 1200;
+constexpr uint32_t gMessageTimeoutMsec    = 1000;
+constexpr size_t kMaxSubMessageCount      = 1;
+constexpr int32_t gMessageIntervalSeconds = 1;
```
please put the new ones on the bottom so it's easier to see what changed

yunhanw-google replied:
Done, thanks

asakhart reviewed (commented):
I think in the header of the comment it's worth noting what the expected public API is and some simple pseudo code for how it will be used for subscriptions. An abbreviated version of what is in the tests should do.

mrjerryjohns commented on src/app/ReadHandler.h:
```diff
@@ -117,17 +124,25 @@ class ReadHandler : public Messaging::ExchangeDelegate
     // is larger than current self vended event number
     void MoveToNextScheduledDirtyPriority();
 
+    bool IsSubscription() { return mInteractionType == InteractionType::Subscribe; }
```
```suggestion
    bool IsSubscriptionType() { return mInteractionType == InteractionType::Subscribe; }
```

mrjerryjohns replied:
Please add symmetric equivalent for read.

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/ReadHandler.h:
```diff
@@ -117,17 +124,25 @@ class ReadHandler : public Messaging::ExchangeDelegate
-        Reporting,         ///< The handler is reporting
+        Uninitialized = 0,     ///< The handler has not been initialized
+        Initialized,           ///< The handler has been initialized and is ready
+        Reportable,            ///< The handler has received read request and is waiting for the data to send to be available
```
Reportable as a term doesn't describe a state, but a capability.  Would recommend the following:

```suggestion
        GeneratingReports,            ///< The handler has received either a Read or Subscribe request and is the process of generating a report.
```

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/ReadHandler.h:
```diff
@@ -117,17 +124,25 @@ class ReadHandler : public Messaging::ExchangeDelegate
+        Uninitialized = 0,     ///< The handler has not been initialized
+        Initialized,           ///< The handler has been initialized and is ready
+        Reportable,            ///< The handler has received read request and is waiting for the data to send to be available
+        Reporting,             ///< The handler is reporting
```
Not clear whether this refers the act of creating a report, or already having sent a report. 

```suggestion
        AwaitingReportResponse,             ///< The handler has sent the report to the client and is awaiting a status response.
```

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/ReadHandler.h:
```diff
@@ -117,17 +124,25 @@ class ReadHandler : public Messaging::ExchangeDelegate
+        Initialized,           ///< The handler has been initialized and is ready
+        Reportable,            ///< The handler has received read request and is waiting for the data to send to be available
+        Reporting,             ///< The handler is reporting
+        SubscribingResponding, ///< The handler is wait for status report for subscribe response
```
The current IM/Encoding specs do not specify a status response to be sent back in response to a Subscribe Response. Please update.

yunhanw-google replied:
i see, thanks

yunhanw-google replied:
updated

mrjerryjohns commented on src/app/reporting/Engine.h:
```diff
@@ -119,6 +121,8 @@ class Engine
      */
     uint32_t mCurReadHandlerIdx = 0;
+
+    ClusterInfo mChangeSetPathList[CHIP_IM_MAX_NUM_SERVER_DIRTY_PATHS];
```
Please add comments here describing what this is, which is a global set of all dirty attributes across the device.

Also, since your above method is named `SetDirty`, this should have the term Dirty in it to align (or conversely, name the above `SetChanged`).

Suggestion:

```suggestion
    ClusterInfo mGlobalDirtyPathList[CHIP_IM_MAX_NUM_SERVER_DIRTY_PATHS];
```

yunhanw-google replied:
Updated code

yunhanw-google replied:
Updated code: ClusterInfo itself has mpNext inside, and there is one clusterInfo pool from interaction model engine, which currently holds the interested path for each active subscribe client, I am extending its responsibility and reuse its functionality to hold changed path as well during post-subscription, every time application call setChange/SetDirty, which would use the existing MergeOverlappedAttributePath and merge overlapped path into mpChangePathList, otherwise, it uses the existing PushFront to push one available clusterInfo into mpChangePathList, regarding MergeOverlappedAttributePath.

mrjerryjohns commented on src/app/reporting/Engine.h:
```diff
@@ -77,6 +77,8 @@ class Engine
      */
     CHIP_ERROR ScheduleRun();
 
+    void SetDirty(ClusterInfo & aClusterInfo);
```
Please add comments indicating what 'dirty' means.

yunhanw-google replied:
updated with comment

mrjerryjohns commented on src/app/reporting/Engine.cpp:
```diff
@@ -109,6 +107,32 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
+                        {
+                            err = RetrieveClusterData(attributeDataList, path);
+                        }
+                        else if (path.IsAttributePathSupersetOf(*clusterInfo))
```
On first glance, checking for supersets in both directions is not trivially understandable. Adding comments as to why you're doing so would really help the reader.

yunhanw-google replied:
Updated code

mrjerryjohns commented on src/app/reporting/Engine.h:
```diff
@@ -119,6 +121,8 @@ class Engine
      */
     uint32_t mCurReadHandlerIdx = 0;
+
+    ClusterInfo mChangeSetPathList[CHIP_IM_MAX_NUM_SERVER_DIRTY_PATHS];
```
I also see this pattern fairly pervasively across the IM when dealing with pooled objects, where the flag for a pooled object being 'in-use' is embedded in the object itself, and done so fairly inconsistently.

In this example here for ClusterInfo, you use the `IsDirty` flag to indicate that that. This is not only a poor name (because it's odd for a dirty path list to _not have a dirty item in it_), but we should really not embed the validity of the object _within the object itself_.

For starters, I'd recommend this move to using the `BitMapObjectPool<T>`, which would permit you to remove the `IsDirty` flag from the ClusterInfo object.

yunhanw-google replied:
just remove this one, and use the existing cluster info pool from interaction model engine, would have seperate pr to update cluster info pool with object pool in interaction model engine,

mrjerryjohns commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +367,65 @@ void Engine::Run()
+                if (clusterInstance->IsAttributePathSupersetOf(aClusterInfo) ||
+                    aClusterInfo.IsAttributePathSupersetOf(*clusterInstance))
+                {
+                    for (auto & path : mChangeSetPathList)
```
Except for this for loop, I don't think the giant outer for loop here starting from 398 to 425 is needed - it's pre-mature optimization to need to mark the paths stored in the ReadHandlers as dirty _in addition to_ the core dirty set tracked in the engine. In your implementation in `BuildSingleReportDataAttributeDataList`, you already do this permuted-path-intersection check, so doing it extra here doesn't buy you much. I'd rather this logic be simpler to read _for starters_, and then optimize if we think we need to later.

The other nice side-effect of removing this optimization is that you can then delete the dirty flag from within `ClusterInfo`, which then pivots it to be purely a data structure that stores a path.

yunhanw-google replied:
agree, will do it, thanks

mrjerryjohns commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +367,65 @@ void Engine::Run()
+                            return;
+                        }
+                    }
+                    ChipLogProgress(DataManagement, "<RE> SetDirty fails with no available clean path in mChangeSetPathList");
```
We need a graceful fall back story, since just returning and not marking the path as dirty isn't going to be cut it, since no reports will get sent indefinitely for that dirty item till it changes again next time.

Not to be addressed here, but needs a follow-up issue.

yunhanw-google replied:
updated.

mrjerryjohns commented on src/app/ReadHandler.cpp:
```diff
@@ -25,14 +25,17 @@
 namespace app {
 CHIP_ERROR ReadHandler::Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate,
-                             Messaging::ExchangeContext * apExchangeContext)
+                             Messaging::ExchangeContext * apExchangeContext, InteractionType aInteractionType)
```
Can't comment in the right section, but just a note that right now, the ReadHandler is generating new exchanges for each report messages, even if all the reports are chunked as part of the same Action. This needs fixing, but can happen later.

yunhanw-google replied:
During subscription establishenment procedure, the initial report is using same exchange context. During post establishment, the report is using new exchange context.

erjiaqing commented on src/app/util/ember-compatibility-functions.cpp:
```diff
@@ -461,5 +461,21 @@ CHIP_ERROR WriteSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVReader & a
                                                   Protocols::SecureChannel::Id, imCode);
 }
 
+void emberAfReportingAttributeChangeCallback(EndpointId endpoint, ClusterId clusterId, AttributeId attributeId, uint8_t mask,
```
This function should be placed in global namespace.

yunhanw-google replied:
I just remove this function, i prefer you keep. this function in your pr, since i don't have test in my pr for this part

mrjerryjohns commented on src/app/InteractionModelEngine.h:
```diff
@@ -141,6 +149,9 @@ class InteractionModelEngine : public Messaging::ExchangeDelegate
 
     void ReleaseClusterInfoList(ClusterInfo *& aClusterInfo);
     CHIP_ERROR PushFront(ClusterInfo *& aClusterInfoLisst, ClusterInfo & aClusterInfo);
+    // Merges aAttributePath inside apAttributePathList if current path is overlapped with existing path in apAttributePathList
```
This needs more comments and a definition of what 'overlap' means.

mrjerryjohns commented on src/app/InteractionModelEngine.h:
```diff
@@ -141,6 +149,9 @@ class InteractionModelEngine : public Messaging::ExchangeDelegate
     CHIP_ERROR PushFront(ClusterInfo *& aClusterInfoLisst, ClusterInfo & aClusterInfo);
+    // Merges aAttributePath inside apAttributePathList if current path is overlapped with existing path in apAttributePathList
+    bool MergeOverlappedAttributePath(ClusterInfo * apAttributePathList, ClusterInfo & aAttributePath);
+    bool IsAttributePathInterested(ClusterInfo & aAttributePath);
```
Did you mean to say 'intersected'?

Also, you use the term 'overlap' above, but use a new term 'intersected'? Would recommend sticking to a single term if possible

```suggestion
            bool IsOverlappedAttributePath(ClusterInfo & aAttributePath);
```

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/reporting/Engine.cpp:
```diff
@@ -336,6 +363,18 @@ void Engine::Run()
+    InteractionModelEngine::GetInstance()->ReleaseClusterInfoList(mpChangePathList);
+}
+
+CHIP_ERROR Engine::SetChange(ClusterInfo & aClusterInfo)
```
I honestly preferred the term  'dirty' before since it's a well understood term in the industry (especially when it comes to caches, and the concept of invalidation).

Since you've moved the dirty flag out of ClusterInfo, it stands to reason this could still use the term 'dirty'.

Otherwise, if you want to stick to 'change', then this should be past-tense:

```suggestion
CHIP_ERROR Engine::MarkChanged(ClusterInfo & aClusterInfo)
```

yunhanw-google replied:
update it with SetDirty, thanks

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -99,17 +97,39 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
         {
-            if (apReadHandler->IsInitialReport())
+            ClusterInfo * path = mpChangePathList;
+            while (path != nullptr)
```
If you use for loop here

yunhanw-google replied:
done, thanks

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -99,17 +97,39 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
+                {
+                    // partial overlap is not possible, hence the 'continue' here: clusterInfo and path have nothing in
+                    // common.
+                    path = path->mpNext;
```
You can remove this line, and the code will be more readable.

yunhanw-google replied:
updated

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -99,17 +97,39 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
                 VerifyOrExit(err == CHIP_NO_ERROR,
                              ChipLogError(DataManagement, "<RE:Run> Error retrieving data from cluster, aborting"));
                 attributeClean = false;
+                path           = path->mpNext;
```
And this will always be called!

yunhanw-google replied:
updated.

erjiaqing commented on src/lib/core/CHIPConfig.h:
```diff
@@ -2439,7 +2439,7 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
  */
 #ifndef CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS
-#define CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS 4
+#define CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS 7
```
Why 7 for default?  Since 7 does not looks like a common random size for a pool.

yunhanw-google replied:
would update it soon.

erjiaqing commented on src/app/reporting/Engine.cpp:
```diff
@@ -99,17 +97,39 @@ CHIP_ERROR Engine::BuildSingleReportDataAttributeDataList(ReportData::Builder &
     // TODO: Need to handle multiple chunk of message
     while (clusterInfo != nullptr)
```
The same for this one, the `cluster = clusterInfo->mpNext` is 30 lines below, and I have to scoll my screen to see it.

yunhanw-google replied:
Done, thanks

mrjerryjohns commented on src/app/ReadHandler.cpp:
```diff
@@ -51,6 +51,7 @@ CHIP_ERROR ReadHandler::Init(Messaging::ExchangeManager * apExchangeMgr, Interac
     MoveToState(HandlerState::Initialized);
     mpDelegate       = apDelegate;
     mSubscriptionId  = 0;
+    mHoldReport = false;
```
I don't see anywhere this flag is actually used to hold off reports?

yunhanw-google committed:
[im/subscribe] Add IM subscription support upon IM read

--Problems:
Follow IM spec chapter 5 and IM encoding spec 7.3 and 7.4 to achieve
subscription functionality.
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Interaction-Model.adoc#5-subscribe-interaction
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Encoding-Specification.adoc#subscriberequestmessage

--Change overview
Establishing distinct subscriptions and tracking their respective configurations from multiple clients simultaneously.
Sharing of reporting logic between reads and subscriptions to maximize re-use.
Multiple paths per subscription
Basic chunking of data across multiple reports.
Automatic liveness assessment on client and server and termination of subscription accordingly.
Eventing support in addition to attribute reporting.
Spec compliance to the encoding specification with TLV-based payloads.
Preservation of the application level APIs for client-side reporting, but with a swap of the internals to use the new machinery

yunhanw-google committed:
add multiple tests

--Testing
Add unit test for positive and negative subscribe test
Add integrated cirque test for subscribe
